src/Application/Contracts/IParticipantRepository.cs

```diff
@@ -9,12 +12,20 @@ public interface IParticipantRepository
         Task<ParticipantDetails> GetParticipantDetailsAsync(string participantId);
         Task<ParticipantDemographics> GetParticipantDemographicsAsync(string participantId);
         Task CreateParticipantDetailsAsync(ParticipantDetails entity);
+        
         Task UpdateParticipantDetailsAsync(ParticipantDetails entity);
         Task CreateParticipantDemographicsAsync(ParticipantDemographics entity);
         Task AddDemographicsToNhsUserAsync(ParticipantDemographics entity, string nhsId);
         Task UpdateParticipantDemographicsAsync(ParticipantDemographics entity);
         Task DeleteParticipantDetailsAsync(ParticipantDetails entity);
+        
         Task CreateAnonymisedDemographicParticipantDataAsync(ParticipantDetails entity);
         Task<ParticipantDetails> QueryIndexForParticipantDetailsAsync(string query, string colName);
+        Task CreateParticipantAsync(Participant entity);
+        IAsyncEnumerable<Participant> GetAllAsync(
```

Comment:
Make `IParticipantRepository` implement `IAsyncEnumerable<Participant>`.
`IParticipantRepository : IAsyncEnumerable<Participant> { }`

The method `GetAllAsync()` is confusing and unnecessary.
You want to enumerate the repository, so you can just do that directly. Also the `...Async` suffix makes it look like you have to await this call directly.

Reply:
I have done this and it made me implement another method.  Still not sure how this is working, might be worth a catch up?  I understand WHAT it is doing just not the how

Reply:
`IAsyncEnumerable` is a 'built in' interface that the compiler looks for when iterating over a collection with `await foreach`.

Under the hood it calls the implementation of `IAsyncEnumerable.GetAsyncEnumerator()` on your class and iterates over that.

That will be the method you were forced to implement and you would just put your implementation of `GetAllAsync()` in there.

Doing it this way makes the intent/behaviour of your class more evident. It says "`IParticipantRepository` is something you can iterate over asynchronously because it implements the `IAsyncEnumerable` interface."

`GetAllAsync()` is something custom to this codebase and is not based on the language conventions so a reader can't infer too much about what it does right away.

In fact it is slightly confusing as it might appear that it should be used like this:
`List<Participant> participants = await _participantRepository.GetAllAsync();`

See:
https://learn.microsoft.com/en-us/archive/msdn-magazine/2019/november/csharp-iterating-with-async-enumerables-in-csharp-8#a-tour-through-async-enumerables